Fix URL rewrite match search when switching store

aapokiiso opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
When "Use Store Code in URLs" is enabled, Magento 2 currently redirects to the homepage if changing the store from a product, category or CMS page. So basically any page which uses URL rewrites. There has been some work done on this by Magento (see issue #2786), but there's no definitive solution yet. This PR aims to fix the issues with the following changes:

- Uses target path to match URL rewrites between
  stores. Request path (used previously) can be
  different between stores.
- Changes the target URL if matching
  URL rewrite is found. Previously it just
  redirected to the homepage in case the rewrite
  wasn't found, and did nothing if found.

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#2786: Store View Switching

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Start with a clean Magento installation.
2. Create a new store view called "Test store" with the code "test" in the "Main Website" website.
3. Enable "Use Store Code in URLs" globally.
4. Create a product with a URL key "product-1" in the "Default Store View" scope.
5. Change the scope to "Test store" in the product editor.
6. Set the product URL key as "product-1-test-store" in the "Test store" scope. Save the product.
7. Navigate to the frontend http://<your-store.com>/default/product-1 (eg. in the default store)
8. Change the store through the store switcher to "Test store". This should go through the stores/store/redirect controller. **Magento will redirect you http://<your-store.com>/test, aka the homepage for "Test store"**. This PR will change it so that the user remains on the product page, eg. http://<your-store.com>/test/product-1-test-store.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

aapokiiso committed:
Fix URL rewrite match search when switching store

Notable changes:
- Uses target path to match URL rewrites between
  stores. Request path (used previously) can be
  different between stores.
- Changes the target URL if matching
  URL rewrite is found. Previously it just
  redirected to the homepage in case the rewrite
  wasn't found, and did nothing if found.

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=20093) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=20093) before we can accept your contribution.<br/>

magento-engcom-team commented:
Hi @aapokiiso. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

davidverholen commented:
@magento-engcom-team give me test instance

magento-engcom-team commented:
Hi @davidverholen. Thank you for your request. I'm working on Magento instance for you

davidverholen commented:
the test seems to fail since url rewriting (index.php to /) is not enabled in the integration test suite.
This leads to the url being generated as http://localhost/index.php/page-c/ instead of the expected http://localhost/page-c

@aapokiiso can you update the respective test case and see if the build succeeds please

aapokiiso committed:
Correct base URL in URL rewrite test case

"/index.php" -> "/" rewrite is not configured
on test servers, so we need to include "index.php"
in test URLs.

aapokiiso committed:
Remove redundant else-clause from store switch

Commit:
add trailing slash to expected url to match generated url

magento-engcom-team commented:
@aapokiiso thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

hostep commented:
Thanks @aapokiiso 
But be aware that there is another PR which attempt to solve the same issue: https://github.com/magento/magento2/pull/19798

Not sure which one of these two is the better one...

aapokiiso commented:
@hostep Nice find, they seem to achieve the same thing with some minor differences. 

My PR selects the first matching URL rewrite, whereas #19798 selects the last one. As I understand, normally there wouldn't be multiple URL rewrites for the same target path + store combination, so this shouldn't matter.

The other difference is the target URL generation based on the rewrite request path. My PR uses `Magento\Store\Model\Store::getUrl`, whereas #19798 combines the store base URL with the request path using PHP string concatenation. Again, this _shouldn't_ make any difference in real life.

davidverholen commented:
@aapokiiso the test is failing again, now having http://localhost/index.php/page-c-on-2nd-store/ as the redirected url.
It seems correct if the fixture has this url set on the second store. Not sure why it was still another url before

aapokiiso committed:
Use string concat to combine base URL with rewrite

Magento\Store\Model\Store::getUrl expects the $route
parameter to be a module/controller/action combination,
not a free-form URL path like in this case.

aapokiiso committed:
Merge branch 'url-rewrite-switch-store-correct-path' of github.com:aapokiiso/magento2 into url-rewrite-switch-store-correct-path

davidverholen commented:
@aapokiiso I think you are right and the test was basically wrong succeeding before since it should have expected the url key form the second store

aapokiiso committed:
Expect 2nd store URL in store switcher test

davidverholen commented:
Hi @aapokiiso,
thank you for your contribution, I'm closing this as a duplicate of https://github.com/magento/magento2/pull/19798 as was stated before.

aapokiiso commented:
@davidverholen Ok, thanks for the info! By the way, the "Contribution Survey" link above is broken.

davidverholen commented:
@aapokiiso thanks for the hint, I'll try find someone where I can forward this to ;)

davidverholen commented:
@aapokiiso did you login on the page via github? Just wondering if it is maybe only available for the PR author

sidolov commented:
Hi @aapokiiso , @davidverholen survey link was fixed, you may try once again. I'm sorry for the inconvenience

hostep commented:
@sidolov: sorry to intrude in this thread, I also recently got such a link in an accepted PR, and now it works indeed. But it asks me to login and then says I have to give Magento read and **write** access to all my public repo's. I'm not really comfortable with this, is there somehow a way to reduce the amount of permissions we must give you to login to https://opensource.magento.com/ ?

Exact permissions it asks are:
> - Magento Community Portal by magento wants to access your {username} account
> - This application will be able to read your private email addresses.
> - This application will be able to read and write all public repository data. This includes the following:
>     * Code
>     * Issues
>     * Pull requests
>     * Wikis
>     * Settings
>     * Webhooks and services
>     * Deploy keys
